gpflow/check_shapes/check_shapes.py

```diff
@@ -0,0 +1,520 @@
+        _consume(_ATTRIBUTE_ARGUMENT_RE)
+        if match:
+            result = _AttributeArgumentRef(result, match.group(1))
+            continue
+        _consume(_INDEX_ARGUMENT_RE)
+        if match:
+            result = _IndexArgumentRef(result, int(match.group(1)))
+            continue
+        assert False, _create_error_message()
+    assert start == len(argument_ref_str), _create_error_message()
+    return result
+
+
+class ArgumentReferenceError(Exception):
+    """ Error raised if the argument to check the shape of could not be resolved. """
+
+    def __init__(self, func: _C, arg_map: Mapping[str, Any], arg_ref: _ArgumentRef) -> None:
+        func_info = _FunctionDebugInfo.create(func)
+        lines = [
+            "Unable to resolve argument / missing argument.",
+            f"    Function: {func_info.name}",
+            f"    Declared: {func_info.path_and_line}",
+            f"    Argument: {arg_ref}",
+        ]
+
+        super().__init__("\n".join(lines))
+
+        self.func = func
+        self.arg_map = arg_map
+        self.arg_ref = arg_ref
+
+
+class ShapeMismatchError(Exception):
+    """ Error raised if a function is called with tensors of the wrong shape. """
+
+    def __init__(
+        self,
+        func: _C,
+        specs: Sequence[_ParsedArgumentSpec],
+        arg_map: Mapping[str, Any],
+    ) -> None:
+        func_info = _FunctionDebugInfo.create(func)
+        lines = [
+            "Tensor shape mismatch in call to function.",
+            f"    Function: {func_info.name}",
+            f"    Declared: {func_info.path_and_line}",
+        ]
+        for spec in specs:
+            actual_shape = spec.argument_ref.get(func, arg_map).shape
+            if isinstance(actual_shape, tf.TensorShape) and actual_shape.rank is None:
+                actual_str = "<Unknown>"
+            else:
+                actual_str = f"({', '.join(str(dim) for dim in actual_shape)})"
+            lines.append(
+                f"    Argument: {spec.argument_ref}, expected: {spec.shape}, actual: {actual_str}"
+            )
+
+        super().__init__("\n".join(lines))
+
+        self.func = func
+        self.specs = specs
+        self.arg_map = arg_map
+
+
+@dataclass
+class _FunctionDebugInfo:
+    """
+    Information about a function, to print in error messages.
+    """
+
+    name: str
+    path_and_line: str
+
+    @staticmethod
+    def create(func: _C) -> "_FunctionDebugInfo":
+        name = func.__qualname__
+        try:
+            path = inspect.getsourcefile(func)
+        except Exception:  # pragma: no cover
+            path = "<unknown file>"
+        try:
+            _, line_int = inspect.getsourcelines(func)
+            line = str(line_int)
+        except Exception:  # pragma: no cover
+            line = "<unknown lines>"
+        path_and_line = f"{path}:{line}"
+
+        return _FunctionDebugInfo(name=name, path_and_line=path_and_line)
+
+
+def inherit_check_shapes(func: _C) -> _C:
```

Comment:
Can we make this the default behaviour and simply use the decorator to overwrite the shape checks?